src/IO/RemoteReadBufferCache.cpp

```diff
@@ -0,0 +1,596 @@
+            }
+        }
+        std::unique_lock lock(mutex);
+        current_offset += total_bytes;
+        download_finished = true;
+        flush();
+        out_file->close();
+        delete out_file;
+        out_file = nullptr;
+        remote_readbuffer = nullptr;
+        lock.unlock();
+        more_data_signal.notify_all();
+        finish_callback(this);
+        LOG_TRACE(
+            &Poco::Logger::get("RemoteCacheController"), "finish download.{} into {}. size:{} ", remote_path, local_path, current_offset);
+    };
+    download_thread->scheduleOrThrow(task);
+}
+
+void RemoteCacheController::flush()
+{
+    if (out_file != nullptr)
+    {
+        out_file->flush();
+    }
+    Poco::JSON::Object jobj;
+    jobj.set("schema", schema);
+    jobj.set("cluster", cluster);
+    jobj.set("remote_path", remote_path);
+    jobj.set("downloaded", download_finished ? "true" : "false");
+    jobj.set("last_mod_ts", last_mod_ts);
+    std::stringstream buf; // STYLE_CHECK_ALLOW_STD_STRING_STREAM
+    jobj.stringify(buf);
+
+    std::ofstream meta_file(local_path + "/meta.txt", std::ios::out);
+    meta_file << buf.str();
+    meta_file.close();
+}
+
+RemoteCacheController::~RemoteCacheController()
+{
+    delete out_file;
+
+    if (download_thread != nullptr)
+    {
+        download_thread->wait();
+        delete download_thread;
+    }
+}
+
+void RemoteCacheController::close()
+{
+    // delete the directory
+    LOG_TRACE(&Poco::Logger::get("RemoteCacheController"), "release local resource " + remote_path + ", " + local_path);
+    Poco::File file(local_path);
+    file.remove(true);
+}
+
+FILE * RemoteCacheController::allocFile(std::string * local_path_)
+{
+    if (download_finished)
+        *local_path_ = local_path + "/data.bin";
+    FILE * fs = fopen((local_path + "/data.bin").c_str(), "r");
+    if (fs == nullptr)
+        return fs;
+    std::lock_guard lock{mutex};
+    opened_file_streams.insert(fs);
+    return fs;
+}
+
+void RemoteCacheController::deallocFile(FILE * fs)
+{
+    {
+        std::lock_guard lock{mutex};
+        auto it = opened_file_streams.find(fs);
+        if (it == opened_file_streams.end())
+        {
+            std::string err = "try to close an invalid file " + remote_path;
+            throw Exception(err, ErrorCodes::BAD_ARGUMENTS);
+        }
+        opened_file_streams.erase(it);
+    }
+    fclose(fs);
+}
+
+LocalCachedFileReader::LocalCachedFileReader(RemoteCacheController * cntrl_, size_t size_)
+    : offset(0), file_size(size_), fs(nullptr), controller(cntrl_)
+{
+    fs = controller->allocFile(&local_path);
+    if (fs == nullptr)
+        throw Exception("alloc file failed.", ErrorCodes::BAD_GET);
+}
+LocalCachedFileReader::~LocalCachedFileReader()
+{
+    controller->deallocFile(fs);
+}
+
+size_t LocalCachedFileReader::read(char * buf, size_t size)
+{
+    auto wret = controller->waitMoreData(offset, offset + size);
+    if (wret < 0)
+        return 0;
```

Comment:
Better to have enum with possible states.

Reply:
done